File: backend/app/services/description_merger.py
# -*- coding: utf-8 -*-
"""DescriptionMerger - left-joins a DESC column onto already-converted rows
by PPID (V1.06 "Add Description" feature).

Intentionally decoupled from ExcelTransformer, same spirit as rule_manager:
this only *adds* a column to data that already exists, it never re-derives
or mutates the converted rows themselves.
"""

from dataclasses import dataclass

import pandas as pd

from app.models.constants import MISSING_VALUE
from app.utils.df_helpers import insert_column_after
from app.utils.excel_io import InvalidExcelFormatError

# Cap on how many duplicate/unmatched PPIDs are echoed back in an error
# message or response - production description files can be large, and
# nobody reads a list of 5,000 PPIDs anyway.
_MAX_LISTED_PPIDS = 50


@dataclass(frozen=True)
class MergeResult:
    df: pd.DataFrame
    matched_count: int
    unmatched_count: int
    unmatched_ppids: list[str]


def merge_description(rows_df: pd.DataFrame, desc_df: pd.DataFrame) -> MergeResult:
    """Left-join `desc_df` (PPID, DESC) onto `rows_df` by PPID.

    - `rows_df` (the already-converted data) is never mutated - a fresh
      copy is returned. Multiple rows sharing a PPID all receive the same
      DESC (Requirement: "DESC shall be added to every matching row").
    - `desc_df`'s PPID must be unique; a duplicate is a validation error,
      not a silently-resolved conflict, since there'd be no principled way
      to pick which DESC wins.
    - Rows whose PPID has no match get `MISSING_VALUE` ("-"), consistent
      with how the rest of the app marks an absent field.
    - matched/unmatched counts are per distinct PPID (not per row), to
      match the Status Bar's "Matched PPIDs" / "Unmatched PPIDs" labels.
    """
    if "PPID" not in rows_df.columns:
        raise InvalidExcelFormatError("Converted data is missing a PPID column.")

    desc_lookup = desc_df.copy()
    desc_lookup["PPID"] = desc_lookup["PPID"].astype(str).str.strip()

    duplicate_mask = desc_lookup["PPID"].duplicated()
    if duplicate_mask.any():
        duplicates = sorted(desc_lookup.loc[duplicate_mask, "PPID"].unique().tolist())
        shown = ", ".join(duplicates[:_MAX_LISTED_PPIDS])
        suffix = f" and {len(duplicates) - _MAX_LISTED_PPIDS} more" if len(duplicates) > _MAX_LISTED_PPIDS else ""
        raise InvalidExcelFormatError(f"Description file has duplicate PPID(s): {shown}{suffix}")

    lookup = dict(zip(desc_lookup["PPID"], desc_lookup["DESC"]))

    ppid_series = rows_df["PPID"].astype(str).str.strip()
    desc_values = ppid_series.map(lookup)
    desc_values = desc_values.where(desc_values.notna(), MISSING_VALUE)

    unique_ppids = ppid_series.unique().tolist()
    matched_ppids = [p for p in unique_ppids if p in lookup]
    unmatched_ppids = sorted(p for p in unique_ppids if p not in lookup)

    # V1.07: DESC always displays/exports immediately after PPID, not at the
    # end - so every consumer (the /api/add-description response, and
    # /api/export once it reads this same column order) gets it right
    # without having to know the rule itself. Shared with /api/export's
    # placement logic via insert_column_after (see app/utils/df_helpers.py).
    merged = insert_column_after(rows_df, "DESC", desc_values.to_numpy(), after="PPID")

    return MergeResult(
        df=merged,
        matched_count=len(matched_ppids),
        unmatched_count=len(unmatched_ppids),
        unmatched_ppids=unmatched_ppids[:_MAX_LISTED_PPIDS],
    )

File: frontend/lib/converter/descriptionMerger.ts
/**
 * DescriptionMerger - TS port of
 * backend/app/services/description_merger.py (V1.10 Browser Edition).
 *
 * Left-joins a DESC column onto already-converted rows by PPID. Kept
 * decoupled from the transformer/rule engine, same spirit as the Python
 * version: this only *adds* a column to data that already exists.
 */

import { MISSING_VALUE } from "./constants";
import { insertFieldAfter } from "./dfHelpers";

// Cap on how many duplicate/unmatched PPIDs are echoed back - production
// description data can be large, and nobody reads a list of 5,000 PPIDs.
const MAX_LISTED_PPIDS = 50;

export class InvalidExcelFormatError extends Error {}

export type DescriptionRow = { PPID: unknown; DESC: unknown };

export type MergeResult = {
  rows: Record<string, unknown>[];
  fields: string[];
  matchedCount: number;
  unmatchedCount: number;
  unmatchedPpids: string[];
};

/**
 * Left-joins `descRows` (PPID, DESC) onto `rowsData` (the already-converted
 * data) by PPID.
 *
 * - `rowsData` is never mutated - every returned row is a fresh object.
 *   Multiple rows sharing a PPID all receive the same DESC.
 * - `descRows`'s PPID must be unique; a duplicate is a validation error,
 *   not a silently-resolved conflict.
 * - Rows whose PPID has no match get MISSING_VALUE ("-").
 * - matched/unmatched counts are per distinct PPID, matching the Status
 *   Bar's "Matched PPIDs" / "Unmatched PPIDs" labels.
 */
export function mergeDescription(rowsData: Record<string, unknown>[], descRows: DescriptionRow[]): MergeResult {
  if (rowsData.length === 0 || !("PPID" in rowsData[0])) {
    throw new InvalidExcelFormatError("Converted data is missing a PPID column.");
  }

  const cleanedDesc = descRows.map((r) => ({ ppid: String(r.PPID ?? "").trim(), desc: r.DESC }));

  const seen = new Set<string>();
  const duplicates = new Set<string>();
  for (const r of cleanedDesc) {
    if (seen.has(r.ppid)) duplicates.add(r.ppid);
    seen.add(r.ppid);
  }
  if (duplicates.size > 0) {
    const sorted = Array.from(duplicates).sort();
    const shown = sorted.slice(0, MAX_LISTED_PPIDS).join(", ");
    const suffix = sorted.length > MAX_LISTED_PPIDS ? ` and ${sorted.length - MAX_LISTED_PPIDS} more` : "";
    throw new InvalidExcelFormatError(`Description file has duplicate PPID(s): ${shown}${suffix}`);
  }

  const lookup = new Map(cleanedDesc.map((r) => [r.ppid, r.desc]));

  const uniquePpids: string[] = [];
  const uniqueSeen = new Set<string>();
  const mergedRows = rowsData.map((row) => {
    const ppid = String(row.PPID ?? "").trim();
    if (!uniqueSeen.has(ppid)) {
      uniqueSeen.add(ppid);
      uniquePpids.push(ppid);
    }
    const desc = lookup.has(ppid) ? lookup.get(ppid) : MISSING_VALUE;
    return { ...row, DESC: desc };
  });

  const matchedPpids = uniquePpids.filter((p) => lookup.has(p));
  const unmatchedPpids = uniquePpids.filter((p) => !lookup.has(p)).sort();

  // V1.07: DESC always displays/exports immediately after PPID - shared
  // with /api/export's placement logic (see engine.ts's exportRows, which
  // uses the same insertFieldAfter helper for the same reason).
  const fields = insertFieldAfter(Object.keys(rowsData[0]), "DESC", "PPID");

  return {
    rows: mergedRows,
    fields,
    matchedCount: matchedPpids.length,
    unmatchedCount: unmatchedPpids.length,
    unmatchedPpids: unmatchedPpids.slice(0, MAX_LISTED_PPIDS),
  };
}
